src/wasm/wasm.cpp

```diff
@@ -362,41 +301,55 @@ static void handleUnreachable(Block* block,
 }
 
 void Block::finalize() {
+  if (list.size() == 0) {
+    type = Type::none;
+    return;
+  }
   if (!name.is()) {
-    if (list.size() > 0) {
-      // nothing branches here, so this is easy
-      // normally the type is the type of the final child
-      type = list.back()->type;
-      // and even if we have an unreachable child somewhere,
-      // we still mark ourselves as having that type,
-      // (block (result i32)
-      //  (return)
-      //  (i32.const 10)
-      // )
-      if (type.isConcrete()) {
-        return;
-      }
-      // if we are unreachable, we are done
-      if (type == Type::unreachable) {
+    // nothing branches here, so this is easy
+    // normally the type is the type of the final child
+    type = list.back()->type;
+    // and even if we have an unreachable child somewhere,
+    // we still mark ourselves as having that type,
+    // (block (result i32)
+    //  (return)
+    //  (i32.const 10)
+    // )
+    if (type.isConcrete()) {
+      return;
+    }
+    // if we are unreachable, we are done
+    if (type == Type::unreachable) {
+      return;
+    }
+    // we may still be unreachable if we have an unreachable
+    // child
+    for (auto* child : list) {
+      if (child->type == Type::unreachable) {
+        type = Type::unreachable;
         return;
       }
-      // we may still be unreachable if we have an unreachable
-      // child
-      for (auto* child : list) {
-        if (child->type == Type::unreachable) {
-          type = Type::unreachable;
-          return;
-        }
-      }
-    } else {
-      type = Type::none;
     }
     return;
   }
 
-  TypeSeeker seeker(this, this->name);
-  type = Type::mergeTypes(seeker.types);
-  handleUnreachable(this);
+  // The default type is according to the value that flows out.
+  type = list.back()->type;
+  BranchUtils::BranchSeeker seeker(this->name);
+  Expression* temp = this;
+  seeker.walk(temp);
+  if (seeker.found) {
+    // Take the branch values into account.
+    if (seeker.valueType != Type::none) {
+      type = Type::getLeastUpperBound(type, seeker.valueType);
+    } else {
+      // No value is sent, but as we have a branch we are not unreachable.
+      type = Type::none;
+    }
+  } else {
+    // There are no branches, so this block may be unreachable.
+    handleUnreachable(this);
```

Comment:
It would be good to pass `breakabilityKnown = true, hasBreak = false` if we know there are no branches.